addon-chrome/ChromeActivityStreams.js

```diff
@@ -0,0 +1,243 @@
+		this._setupActionListeners();
+    this._setupChromeListeners();
+	}
+
+  _setupActionListeners() {
+    window.addEventListener(CONTENT_TO_ADDON, (event) => {
+      const action = JSON.parse(event.detail);
+      switch (action.type) {
+        case "TOP_FRECENT_SITES_REQUEST":
+          this._topFrecentSites(action);
+          break;
+        case "RECENT_BOOKMARKS_REQUEST":
+          this._recentBookmarks(action);
+          break;
+        case "RECENT_LINKS_REQUEST":
+          this._recentLinks(action);
+          break;
+        case "HIGHLIGHTS_LINKS_REQUEST":
+          this._highlightsLinks(action);
+          break;
+        case "NOTIFY_HISTORY_DELETE":
+          this._historyDelete(action);
+          break;
+        case "NOTIFY_BOOKMARK_ADD":
+          this._bookmarkAdd(action);
+          break;
+        case "NOTIFY_BOOKMARK_DELETE":
+          this._bookmarkDelete(action);
+          break;
+        case "NOTIFY_BLOCK_URL":
+          this._blockUrl(action);
+          break;
+        case "NOTIFY_UNBLOCK_ALL":
+          this._unblockAll(action);
+          break;
+        case "NOTIFY_OPEN_WINDOW":
+          this._openNewWindow(action);
+          break;
+        case "SEARCH_SUGGESTIONS_REQUEST":
+          this._searchSuggestions(action);
+          break;
+        case "NOTIFY_PERFORM_SEARCH":
+          this._performSearch(action);
+          break;
+        case "SEARCH_STATE_REQUEST":
+          this._searchState();
+          break;
+        case "SEARCH_UISTRINGS_REQUEST":
+          this._searchUIStrings();
+          break;
+      }
+    }, false);
+  }
+
+  _setupChromeListeners() {
+    chrome.history.onVisited.addListener((result) => {
+      const row = ChromePlacesProvider.transformHistory(result);
+      ChromePlacesProvider.addHistory(row).then((result) => {
+        ChromePlacesProvider.getHistory().then((histories) => {
+          dispatch({
+            type: "RECENT_LINKS_RESPONSE",
+            data: histories
+          });
+        });
+      });
+    });
+
+    chrome.history.onVisitRemoved.addListener((result) => {
+      result.urls.forEach((url) => {
+        ChromePlacesProvider.removeHistory(url);
+        dispatch({
+          type: "NOTIFY_HISTORY_DELETE",
+          data: url
+        });
+      });
+    });
+
+    chrome.bookmarks.onCreated.addListener((id, result) => {
+      const isFolder = !result.url;
+      if (isFolder) return;
+      const row = ChromePlacesProvider.transformBookmark(result);
+      ChromePlacesProvider.addBookmark(row);
+      dispatch({
+        type: "RECENT_BOOKMARKS_RESPONSE",
+        data: [row],
+        meta: {prepend: true}
+      });
+    });
+
+    chrome.bookmarks.onRemoved.addListener((result) => {
+      ChromePlacesProvider.removeBookmark(result);
+      dispatch({
+        type: "NOTIFY_BOOKMARK_DELETE",
+        data: result
+      });
+    });
+  }
+
+  _topFrecentSites(action) {
+    ChromePlacesProvider.getHistory().then((histories) => {
```

Comment:
frecency stuff should probably be in the model layer (places provider) not in the controller layer (here). You controllers should be pretty light: get data from from the underlying models, dispatch it.